qualtran/bloqs/multiplexers/apply_lth_bloq.py

```diff
@@ -0,0 +1,127 @@
+import numpy as np
+from attrs import frozen
+from numpy.typing import NDArray
+
+from qualtran import Bloq, bloq_example, BloqDocSpec, BoundedQUInt, QBit, Register, Side
+from qualtran._infra.gate_with_registers import merge_qubits, SpecializedSingleQubitControlledGate
+from qualtran.bloqs.multiplexers.unary_iteration_bloq import UnaryIterationGate
+from qualtran.resource_counting import BloqCountT, SympySymbolAllocator
+from qualtran.symbolics import ceil, log2, prod
+
+
+@frozen
+class ApplyLthBloq(UnaryIterationGate, SpecializedSingleQubitControlledGate):  # type: ignore[misc]
+    r"""A SELECT operation that executes one of a list of bloqs $U_l$ based on a quantum index:
+
+    $$
+    \mathrm{SELECT} = \sum_{l}|l \rangle \langle l| \otimes U_l
+    $$
+
+    This bloq uses the unary iteration scheme to apply `ops[selection]` controlled on an optional single-bit `control` register.
+
+    Args:
+        ops: Callable that returns the (i, j, ...)-th bloq to select from, when given (i, j, ...).
+             Each bloq must have identical registers that are all THRU.
+        selection_regs: List of selection registers, 1D selection index inferred if not provided.
+        control_val: If provided, a singly controlled gate is constructed.
+
+    Registers:
+        selection: The index of the bloq in `ops` to execute.
+        control: The control bit if specified above.
+        [user_spec]: The output registers of the bloqs in `ops`.
+
+    References:
+        [Encoding Electronic Spectra in Quantum Circuits with Linear T Complexity](https://arxiv.org/abs/1805.03662).
+        Babbush et. al. (2018). Section III.A. and Figure 7.
+    """
+
+    ops: Callable[..., Bloq]
+    n: Tuple[int, ...]
+    selection_regs: Optional[Tuple[Register, ...]] = None
+    control_val: Optional[int] = None
+
+    @cached_property
+    def _ops(self) -> NDArray[Bloq]:  # type: ignore[type-var]
+        return np.fromfunction(np.vectorize(self.ops), self.n, dtype=object)
+
+    def __attrs_post_init__(self):
+        if prod(self._ops.shape) == 0:
+            raise ValueError("Must have at least one operation.")
+        if not all(u.signature == self._ops.flat[0].signature for u in self._ops.flat):
+            raise ValueError("All ops must have the same signature.")
+        if not all(r.side == Side.THRU for u in self._ops.flat for r in u.signature):
+            raise ValueError("All ops must have only THRU registers.")
+
+    @cached_property
+    def control_registers(self) -> Tuple[Register, ...]:
+        return () if self.control_val is None else (Register('control', QBit()),)
+
+    @cached_property
+    def selection_registers(self) -> Tuple[Register, ...]:
+        if self.selection_regs is None:
+            return (
+                Register(
+                    'selection',
+                    BoundedQUInt(int(ceil(log2(len(self._ops.flat)))), len(self._ops.flat)),
+                ),
+            )
+        return self.selection_regs
```

Comment:
My original suggestion to expect an `NDArray` of `Bloq`s instead of a flat list was so that we can use multiple selection registers instead of a single one and have a more accurate representation of a  "nested for-loop" to index into the NDArray.

Thus, we should have `len(self._ops.shape)` different selection registers instead of a single selection register of of length `self._ops.flat`.

Reply:
That makes sense. I'll make the change.